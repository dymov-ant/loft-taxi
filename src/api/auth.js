import { api } from './index';

export const registration = (payload) => api.post('/register', payload);

export const login = (payload) => api.post('/auth', payload);
