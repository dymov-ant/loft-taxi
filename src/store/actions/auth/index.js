import {
  FETCH_LOGIN,
  FETCH_REGISTRATION,
  LOGOUT,
  SET_ERROR,
  SET_IS_AUTH,
  SET_IS_LOADING,
} from '../../reducers/auth/types';

export const authActions = {
  setIsAuth: (payload) => ({ type: SET_IS_AUTH, payload }),
  setIsLoading: (payload) => ({ type: SET_IS_LOADING, payload }),
  setError: (payload) => ({ type: SET_ERROR, payload }),
  fetchLogin: (payload) => ({ type: FETCH_LOGIN, payload }),
  fetchRegistration: (payload) => ({ type: FETCH_REGISTRATION, payload }),
  logout: () => ({ type: LOGOUT }),
};
