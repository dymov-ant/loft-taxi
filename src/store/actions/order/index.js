import {
  CLEAR_ROUTE,
  FETCHING_ADDRESS_LIST,
  FETCHING_ERROR_ADDRESS_LIST,
  FETCHING_ERROR_ROUTE,
  FETCHING_ROUTE,
  FETCHING_SUCCESS_ADDRESS_LIST,
  FETCHING_SUCCESS_ROUTE,
  GET_ADDRESS_LIST,
  GET_ROUTE,
} from '../../reducers/order/types';

export const orderActions = {
  fetchingAddressList: () => ({ type: FETCHING_ADDRESS_LIST }),
  fetchingSuccessAddressList: (addressList) => ({ type: FETCHING_SUCCESS_ADDRESS_LIST, payload: addressList }),
  fetchingErrorAddressList: (error) => ({ type: FETCHING_ERROR_ADDRESS_LIST, payload: error }),
  getAddressList: () => ({ type: GET_ADDRESS_LIST }),
  fetchingRoute: () => ({ type: FETCHING_ROUTE }),
  fetchingSuccessRoute: (route) => ({ type: FETCHING_SUCCESS_ROUTE, payload: route }),
  fetchingErrorRoute: (error) => ({ type: FETCHING_ERROR_ROUTE, payload: error }),
  clearRoute: () => ({ type: CLEAR_ROUTE }),
  getRoute: (addressFrom, addressTo) => ({ type: GET_ROUTE, payload: { addressFrom, addressTo } }),
};
