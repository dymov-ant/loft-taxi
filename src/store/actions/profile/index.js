import {
  FETCHING_ERROR_PROFILE,
  FETCHING_PROFILE,
  FETCHING_SUCCESS_PROFILE,
  GET_PROFILE,
  IS_SUCCESS_SUBMITTING,
  SET_PROFILE
} from "../../reducers/profile/types"

export const profileActions = {
  fetchingProfile: () => ({type: FETCHING_PROFILE}),
  fetchingSuccess: (card) => ({type: FETCHING_SUCCESS_PROFILE, payload: card}),
  fetchingError: (error) => ({type: FETCHING_ERROR_PROFILE, payload: error}),
  successSubmitting: (isSubmitting) => ({type: IS_SUCCESS_SUBMITTING, payload: isSubmitting}),
  getProfile: () => ({type: GET_PROFILE}),
  setProfile: (cardData) => ({type: SET_PROFILE, payload: cardData})
}