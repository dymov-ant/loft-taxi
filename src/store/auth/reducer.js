import { LOG_IN, LOG_OUT } from "./types"

const initialState = {
  isLoggedIn: false
}

export default function authReducer(state = initialState, action) {
  switch (action.type) {
    case LOG_IN:
      return { ...state, isLoggedIn: true }
    case LOG_OUT:
      return { ...state, isLoggedIn: false }
    default:
      return state
  }
}