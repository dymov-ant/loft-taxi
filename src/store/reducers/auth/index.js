import { SET_ERROR, SET_IS_AUTH, SET_IS_LOADING } from './types';

const initialState = {
  isLoggedIn: false,
  isLoading: false,
  error: null,
};

export default function authReducer(state = initialState, action) {
  switch (action.type) {
    case SET_IS_AUTH:
      return { ...state, isLoggedIn: action.payload };
    case SET_IS_LOADING:
      return { ...state, isLoading: action.payload };
    case SET_ERROR:
      return { ...state, error: action.payload };
    default:
      return state;
  }
}
