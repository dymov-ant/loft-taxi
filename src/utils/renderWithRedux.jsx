import React from 'react';
import { render } from '@testing-library/react';
import { Provider } from 'react-redux';
import { Router } from 'react-router-dom';
// eslint-disable-next-line import/no-extraneous-dependencies
import { createMemoryHistory } from 'history';
// eslint-disable-next-line import/no-extraneous-dependencies
import configureStore from 'redux-mock-store';

const history = createMemoryHistory();
const mockStore = configureStore();

export const renderWithRedux = (component, initialState = {}) => {
  const store = mockStore(initialState);
  return render(
    <Provider store={store}>
      <Router history={history}>{component}</Router>
    </Provider>,
  );
};
