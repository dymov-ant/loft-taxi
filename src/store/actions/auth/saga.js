import { call, put, takeEvery } from 'redux-saga/effects';
import { login, registration } from '../../../api/auth';
import { FETCH_LOGIN, FETCH_REGISTRATION, LOGOUT } from '../../reducers/auth/types';
import { authActions } from './index';
import { AUTH_TOKEN } from '../../../constants';
import { profileActions } from '../profile';
import { orderActions } from '../order';

function* loginWorker(action) {
  try {
    const { payload } = action;
    yield put(authActions.setIsLoading(true));
    const response = yield call(login, payload);
    if (response.data.success) {
      localStorage.setItem(AUTH_TOKEN, response.data.token);
      yield put(authActions.setIsAuth(true));
      yield put(authActions.setError(null));
      yield put(profileActions.getProfile());
    } else {
      yield put(authActions.setError(response.data.error));
      console.log('Ошибка при логине');
    }
  } catch (e) {
    console.log('Ошибка при запросе:', e);
  } finally {
    yield put(authActions.setIsLoading(false));
  }
}

function* registrationWorker(action) {
  try {
    const { payload } = action;
    yield put(authActions.setIsLoading(true));
    const response = yield call(registration, payload);
    if (response.data.success) {
      localStorage.setItem(AUTH_TOKEN, response.data.token);
      yield put(authActions.setIsAuth(true));
      yield put(authActions.setError(null));
    } else {
      yield put(authActions.setError(response.data.error));
      console.log('Ошибка при регистрации');
    }
  } catch (e) {
    console.log('Ошибка при запросе:', e);
  } finally {
    yield put(authActions.setIsLoading(false));
  }
}

function* logoutWorker() {
  localStorage.removeItem(AUTH_TOKEN);
  yield put(authActions.setIsAuth(false));
  yield put(profileActions.fetchingSuccess(null));
  yield put(orderActions.fetchingSuccessRoute([]));
}

export function* authWatcher() {
  yield takeEvery(FETCH_LOGIN, loginWorker);
  yield takeEvery(FETCH_REGISTRATION, registrationWorker);
  yield takeEvery(LOGOUT, logoutWorker);
}
