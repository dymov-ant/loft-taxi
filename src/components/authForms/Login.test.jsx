import React from 'react';
import { fireEvent, waitFor } from '@testing-library/react';
import { renderWithRedux } from '../../utils/renderWithRedux';
import LoginForm from './Login';
import { authActions } from '../../store/actions/auth';

jest.mock('../Spinner', () => () => <div>Spinner</div>);

describe('LoginForm component', () => {
  let initialState;
  beforeEach(() => {
    initialState = {
      auth: {
        isLoading: false,
        error: null,
      },
    };
  });

  it('LoginForm renders', () => {
    const { container } = renderWithRedux(<LoginForm />, initialState);
    expect(container).toBeInTheDocument();
  });

  it('submit login form worked', async () => {
    const fetchLoginSpy = jest.spyOn(authActions, 'fetchLogin');
    const { getByRole, getByTestId } = renderWithRedux(<LoginForm />, initialState);
    const email = getByTestId('email').querySelector('input');
    const password = getByTestId('password').querySelector('input');
    fireEvent.change(email, {
      target: { value: 'test@test.com' },
    });
    fireEvent.change(password, {
      target: { value: '123123' },
    });
    fireEvent.click(getByRole('button'));

    await waitFor(() => expect(fetchLoginSpy).toBeCalled());
    fetchLoginSpy.mockClear();
  });

  it('LoginForm Error render', () => {
    initialState.auth.error = 'Произошла ошибка';
    const { getByText } = renderWithRedux(<LoginForm />, initialState);

    expect(getByText(/Произошла ошибка/i)).toBeInTheDocument();
  });

  it('LoginForm Spinner render', () => {
    initialState.auth.isLoading = true;
    const { getByText } = renderWithRedux(<LoginForm />, initialState);

    expect(getByText(/spinner/i)).toBeInTheDocument();
  });

  it('LoginForm snapshot', () => {
    const loginForm = renderWithRedux(<LoginForm />, initialState);

    expect(loginForm).toMatchSnapshot();
  });
});
