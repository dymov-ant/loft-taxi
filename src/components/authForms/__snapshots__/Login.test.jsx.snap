// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoginForm component LoginForm snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <form
        class="form"
        data-testid="loginForm"
        novalidate=""
      >
        <h2
          class="form__title"
        >
          Войти
        </h2>
        <div
          class="MuiFormControl-root MuiTextField-root MuiFormControl-marginNormal MuiFormControl-fullWidth"
          data-testid="email"
        >
          <label
            class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated"
            data-shrink="false"
          >
            Email
          </label>
          <div
            class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-fullWidth MuiInput-fullWidth MuiInputBase-formControl MuiInput-formControl"
          >
            <input
              aria-invalid="false"
              class="MuiInputBase-input MuiInput-input"
              name="email"
              type="email"
              value=""
            />
          </div>
        </div>
        <div
          class="MuiFormControl-root MuiTextField-root MuiFormControl-marginNormal MuiFormControl-fullWidth"
          data-testid="password"
        >
          <label
            class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated"
            data-shrink="false"
          >
            Пароль
          </label>
          <div
            class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-fullWidth MuiInput-fullWidth MuiInputBase-formControl MuiInput-formControl"
          >
            <input
              aria-invalid="false"
              class="MuiInputBase-input MuiInput-input"
              name="password"
              type="password"
              value=""
            />
          </div>
        </div>
        <a
          class="form__passwordRefresh"
          href="/login"
        >
          Забыли пароль?
        </a>
        <button
          class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-fullWidth"
          tabindex="0"
          type="submit"
        >
          <span
            class="MuiButton-label"
          >
            Войти
          </span>
          <span
            class="MuiTouchRipple-root"
          />
        </button>
        <p
          class="form__text"
        >
          Новый пользователь?
          <a
            href="/registration"
          >
             Регистрация
          </a>
        </p>
      </form>
    </div>
  </body>,
  "container": <div>
    <form
      class="form"
      data-testid="loginForm"
      novalidate=""
    >
      <h2
        class="form__title"
      >
        Войти
      </h2>
      <div
        class="MuiFormControl-root MuiTextField-root MuiFormControl-marginNormal MuiFormControl-fullWidth"
        data-testid="email"
      >
        <label
          class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated"
          data-shrink="false"
        >
          Email
        </label>
        <div
          class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-fullWidth MuiInput-fullWidth MuiInputBase-formControl MuiInput-formControl"
        >
          <input
            aria-invalid="false"
            class="MuiInputBase-input MuiInput-input"
            name="email"
            type="email"
            value=""
          />
        </div>
      </div>
      <div
        class="MuiFormControl-root MuiTextField-root MuiFormControl-marginNormal MuiFormControl-fullWidth"
        data-testid="password"
      >
        <label
          class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated"
          data-shrink="false"
        >
          Пароль
        </label>
        <div
          class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-fullWidth MuiInput-fullWidth MuiInputBase-formControl MuiInput-formControl"
        >
          <input
            aria-invalid="false"
            class="MuiInputBase-input MuiInput-input"
            name="password"
            type="password"
            value=""
          />
        </div>
      </div>
      <a
        class="form__passwordRefresh"
        href="/login"
      >
        Забыли пароль?
      </a>
      <button
        class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-fullWidth"
        tabindex="0"
        type="submit"
      >
        <span
          class="MuiButton-label"
        >
          Войти
        </span>
        <span
          class="MuiTouchRipple-root"
        />
      </button>
      <p
        class="form__text"
      >
        Новый пользователь?
        <a
          href="/registration"
        >
           Регистрация
        </a>
      </p>
    </form>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
